firedrake/slate/preconditioners.py

```diff
@@ -94,33 +96,21 @@ def initialize(self, pc):
         self.unbroken_solution = Function(V)
         self.unbroken_residual = Function(V)
 
-        # Set up the KSP for the hdiv residual projection
-        hdiv_mass_ksp = PETSc.KSP().create(comm=pc.comm)
-        hdiv_mass_ksp.setOptionsPrefix(prefix + "hdiv_residual_")
-
-        # HDiv mass operator
-        p = TrialFunction(V[self.vidx])
-        q = TestFunction(V[self.vidx])
-        mass = ufl.dot(p, q)*ufl.dx
-        # TODO: Bcs?
-        M = assemble(mass, bcs=None, form_compiler_parameters=self.cxt.fc_params)
-        M.force_evaluation()
-        Mmat = M.petscmat
-
-        hdiv_mass_ksp.setOperators(Mmat)
-        hdiv_mass_ksp.setUp()
-        hdiv_mass_ksp.setFromOptions()
-        self.hdiv_mass_ksp = hdiv_mass_ksp
-
-        # Storing the result of A.inv * r, where A is the HDiv
-        # mass matrix and r is the HDiv residual
-        self._primal_r = Function(V[self.vidx])
-
-        tau = TestFunction(V_d[self.vidx])
-        self._assemble_broken_r = create_assembly_callable(
-            ufl.dot(self._primal_r, tau)*ufl.dx,
-            tensor=self.broken_residual.split()[self.vidx],
-            form_compiler_parameters=self.cxt.fc_params)
+        shapes = (V[self.vidx].finat_element.space_dimension(),
+                  np.prod(V[self.vidx].shape))
+        weight_kernel = """
+        for (int i=0; i<%d; ++i) {
+        for (int j=0; j<%d; ++j) {
+        w[i][j] += 1.0;
+        }}""" % shapes
+
+        self.weight = Function(V[self.vidx])
+        par_loop(weight_kernel, ufl.dx, {"w": (self.weight, INC)})
+        self.average_kernel = """
+        for (int i=0; i<%d; ++i) {
+        for (int j=0; j<%d; ++j) {
+        vec_out[i][j] += vec_in[i][j]/w[i][j];
```

Comment:
Since `vec_out` is in a discontinuous space, this is the same as an assignment (each output dof will only be touched once, no?)